src/perform_step/bdf_perform_step.jl

```diff
@@ -262,20 +256,129 @@ function perform_step!(integrator,cache::QNDF1Cache,repeat_step=false)
   fail_convergence && return
   @. u = tmp + γ*z
 
-  if integrator.opts.adaptive
-    D2!(u,uprev,k,cache)
-    @. utilde = (κ*γₖ + inv(k+1)) * D2[1,1]
+  if integrator.opts.adaptive && integrator.success_iter > 0
+    @. D2[1] = u - uprev
+    @. D2[2] = D2[1] - D[1]
+    @. utilde = (κ*γ₁ + inv(k+1)) * D2[2]
     calculate_residuals!(atmp, utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
     integrator.EEst = integrator.opts.internalnorm(atmp)
     if integrator.EEst > one(integrator.EEst)
-      cache.D .= temp_D
-      cache.uprev2 .= temp_u
       return
     end
+  else
+    integrator.EEst = one(integrator.EEst)
   end
   cache.dtₙ₋₁ = dt
   cache.uprev2 .= uprev
   cache.ηold = η
   cache.newton_iters = iter
   f(integrator.fsallast, u, p, t+dt)
 end
+
+function initialize!(integrator, cache::QNDF2ConstantCache)
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+function perform_step!(integrator,cache::QNDF2ConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack pass1,pass2,pdt,uprev2,uprev3,dtₙ₋₁,dtₙ₋₂,D,D2,R,U = cache
+  cnt = integrator.iter
+  pdt = dt
+  k = 2
+  flag = dtₙ₋₁ != dtₙ₋₂
+
+  if flag && cnt != 1
```

Comment:
here, I have tried to changes `dt` forcefully, this is what I have done -> when prev two step size are not same that means we can't use `qndf2`  so first I tried `bdf1` with last step size so that if this gets success, we can apply `qndf2` in next iteration but if it gets failed try with actual step size that was in last iteration.

Reply:
oh yes, that would cause this error. Instead, follow the description described in the paper you found and we should be good. Just calculate the `est`, `est_k-1`, etc.

Reply:
We have to apply `bdf1` in case we can't apply `qndf`, so should I apply `bdf1` until we get the required number of previous constant steps because trick of stepping with the last step in hope that it gets success and would use `qndf` in next iteration, could cause problem (as you mentioned)?
also second paper says -> 
The ode15s is not allowed to change either the order or the step size until a minimum of
(k + 2) consecutive steps are given using the same order formula and the same step size

Reply:
The paper is showing that ode15s isn't doing as quasi-constant as I thought. Instead, think of it as a normal BDF and implement it that way. We can tag the quasi-constant part on later, which is just that if `<qsteady_max` then keep `dt` constant, and if `dt` and the order are constant then re-use the Jacobian. In all other cases it's fine to change `dt` around.